File: Leetcode/TypeScript/876. Middle of the Linked List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function middleNode(head: ListNode | null): ListNode | null {
    let cnt: number = 0;
    let cur: ListNode = head;
    while(cur) {
        cnt++;
        cur = cur.next;
    }
    
    cnt = Math.floor(cnt/2) + 1;
    cur = head;
    
    while(--cnt) {
        cur = cur.next;
    }
    
    return cur;
};

File: Leetcode/Rust/876. Middle of the Linked List.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut s, mut f) = (&head, &head);
        
        while f.is_some() && f.as_ref().unwrap().next.is_some() {
            s = &s.as_ref().unwrap().next;
            f = &f.as_ref().unwrap().next.as_ref().unwrap().next;
        }
        
        s.clone()
    }
}
